src/OrleansCodeGenerator/SerializerGenerator.cs

```diff
@@ -705,12 +705,31 @@ public ExpressionSyntax GetGetter(ExpressionSyntax instance, bool forceAvoidCopy
                     return getValueExpression;
                 }
 
+                ExpressionSyntax deepCopyValueExpression;
+
+                // Addressable arguments must be converted to references before passing.
+                if (typeof(IAddressable).IsAssignableFrom(this.FieldInfo.FieldType)
+                    &&
+                    (typeof(Grain).IsAssignableFrom(this.FieldInfo.FieldType) ||
```

Comment:
If the field is a subclass of `Grain`, then we should not generate code which will assign a `GrainReference` to it. If the field itself is a `Grain`, then we're dead in the water.

We should only do this if the field type is an `interface` and *not* if it is a `Grain`

Of course, the emitted code below, where you perform the `copy.Field = input.Field is Grain ? input.Field.AsReference<T>() : input.Field`

Also worth noting that you're accessing the input field multiple times. Note that `getValueExpression` can emit a method call in the event that the field is private.